From cc39174172eb9c44506d3ccf28294e2c6cac642d Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Fri, 28 Jul 2023 03:34:04 -0700
Subject: [PATCH] Enable Advanced-Route. And Use Compilation Option to
 Enable/Disable it. 	modified:   meson.build 	modified:  
 meson_options.txt 	modified:   src/ipaddress.cpp 	modified:  
 src/ipaddress.hpp 	modified:   src/network_manager.cpp 	modified:  
 src/network_manager.hpp

---
 meson.build             |  1 +
 meson_options.txt       |  2 +-
 src/ipaddress.cpp       | 22 ++++++++++++++++++++++
 src/ipaddress.hpp       |  2 +-
 src/network_manager.cpp | 31 +++++++++++++++++++++++++++++++
 src/network_manager.hpp | 17 ++++++++++++++++-
 6 files changed, 72 insertions(+), 3 deletions(-)

diff --git a/meson.build b/meson.build
index db31e99..f83721b 100644
--- a/meson.build
+++ b/meson.build
@@ -22,6 +22,7 @@ conf_data.set(
 conf_data.set('SYNC_MAC_FROM_INVENTORY', get_option('sync-mac'))
 conf_data.set('PERSIST_MAC', get_option('persist-mac'))
 conf_data.set10('FORCE_SYNC_MAC_FROM_INVENTORY', get_option('force-sync-mac'))
+conf_data.set('AMI_IP_ADVANCED_ROUTING_SUPPORT', get_option('enable-advanced-route'))
 
 sdbusplus_dep = dependency('sdbusplus')
 sdbusplusplus_prog = find_program('sdbus++', native: true)
diff --git a/meson_options.txt b/meson_options.txt
index 735335b..d637f00 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -14,4 +14,4 @@ option('persist-mac', type: 'boolean',
        description: 'Permit the MAC address to be written to the systemd.network config')
 option('force-sync-mac', type: 'boolean',
        description: 'Force sync mac address no matter is first boot or not')
-
+option('enable-advanced-route', type: 'boolean', description: 'Enable advanced route or not')
diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index c8ac522..3438634 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -432,6 +432,24 @@ ObjectPath EthernetInterface::neighbor(std::string ipAddress,
             *addr, std::make_unique<Neighbor>(
                        bus, std::string_view(objPath), *this, *addr, *lladdr,
                        prefixLength, Neighbor::State::Permanent)));
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+        manager.get().addReloadPostHook([&]() {
+            stdplus::In4Addr* inaddr =
+                std::get_if<stdplus::In4Addr>(&(addr.value()));
+            if (inaddr != nullptr)
+            {
+                execute("/usr/bin/ipv4-advanced-route.sh",
+                        "ipv4-advanced-route.sh", interfaceName().c_str(),
+                        "UP");
+            }
+            else
+            {
+                execute("/usr/bin/ipv6-advanced-route.sh",
+                        "ipv6-advanced-route.sh", interfaceName().c_str(),
+                        "UP");
+            }
+        });
+#endif
     }
     else
     {
diff --git a/src/ipaddress.cpp b/src/ipaddress.cpp
index 189c3bf..d78e9fa 100644
--- a/src/ipaddress.cpp
+++ b/src/ipaddress.cpp
@@ -66,7 +66,42 @@ IPAddress::IPAddress(sdbusplus::bus_t& bus,
              true);
     IP::origin(origin, true);
     emit_object_added();
+
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    if (type() == IP::Protocol::IPv4 &&
+        IP::origin() != IP::AddressOrigin::LinkLocal)
+    {
+        execute("/usr/bin/ipv4-advanced-route.sh", "ipv4-advanced-route.sh",
+                parent.get().interfaceName().c_str(), "UP");
+    }
+    else if (type() == IP::Protocol::IPv6 &&
+             IP::origin() != IP::AddressOrigin::LinkLocal)
+    {
+        execute("/usr/bin/ipv6-advanced-route.sh", "ipv6-advanced-route.sh",
+                parent.get().interfaceName().c_str(), "UP");
+    }
+
+#endif
+}
+
+IPAddress::~IPAddress()
+{
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    if (type() == IP::Protocol::IPv4 &&
+        IP::origin() != IP::AddressOrigin::LinkLocal)
+    {
+        execute("/usr/bin/ipv4-advanced-route.sh", "ipv4-advanced-route.sh",
+                parent.get().interfaceName().c_str(), "DOWN");
+    }
+    else if (type() == IP::Protocol::IPv6 &&
+             IP::origin() != IP::AddressOrigin::LinkLocal)
+    {
+        execute("/usr/bin/ipv6-advanced-route.sh", "ipv6-advanced-route.sh",
+                parent.get().interfaceName().c_str(), "DOWN");
+    }
+#endif
 }
+
 std::string IPAddress::address(std::string /*ipAddress*/)
 {
     elog<NotAllowed>(Reason("Property update is not allowed"));
diff --git a/src/ipaddress.hpp b/src/ipaddress.hpp
index 13ab94f..2b17fff 100644
--- a/src/ipaddress.hpp
+++ b/src/ipaddress.hpp
@@ -43,6 +43,8 @@ class IPAddress : public IPIfaces
               stdplus::PinnedRef<EthernetInterface> parent,
               stdplus::SubnetAny addr, IP::AddressOrigin origin);
 
+    ~IPAddress();
+
     std::string address(std::string ipAddress) override;
     uint8_t prefixLength(uint8_t) override;
     std::string gateway(std::string gateway) override;
diff --git a/src/neighbor.cpp b/src/neighbor.cpp
index 4dbb81f..a2d10db 100644
--- a/src/neighbor.cpp
+++ b/src/neighbor.cpp
@@ -61,6 +61,19 @@ void Neighbor::delete_()
 
     parent.get().writeConfigurationFile();
     parent.get().manager.get().reloadConfigs();
+
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    if (auto ipv4 = this->ipAddress().find("."); ipv4 != std::string::npos)
+    {
+        execute("/usr/bin/ipv4-advanced-route.sh", "ipv4-advanced-route.sh",
+                parent.get().interfaceName().c_str(), "DOWN");
+    }
+    else
+    {
+        execute("/usr/bin/ipv6-advanced-route.sh", "ipv6-advanced-route.sh",
+                parent.get().interfaceName().c_str(), "DOWN");
+    }
+#endif
 }
 
 using sdbusplus::xyz::openbmc_project::Common::Error::NotAllowed;
diff --git a/src/network_manager.cpp b/src/network_manager.cpp
index aee4910..d2dc269 100644
--- a/src/network_manager.cpp
+++ b/src/network_manager.cpp
@@ -120,6 +120,10 @@ Manager::Manager(stdplus::PinnedRef<sdbusplus::bus_t> bus,
                            "ERROR", ex);
             }
         }
+
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+        self.get().advanced_route_cond_var.notify_one();
+#endif
         self.get().reloadPostHooks.clear();
     });
     std::vector<
@@ -159,6 +163,11 @@ Manager::Manager(stdplus::PinnedRef<sdbusplus::bus_t> bus,
     systemConf = std::make_unique<phosphor::network::SystemConfiguration>(
         bus, (this->objPath / "config").str);
     setConfDir(confDir);
+
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    advanced_route_lock = std::unique_lock(advanced_route_mutex);
+    advanced_route_worker = std::thread(&Manager::AdvancedRoute, this);
+#endif
 }
 
 void Manager::createInterface(const AllIntfInfo& info, bool enabled)
@@ -645,5 +654,33 @@ void Manager::setConfDir(const fs::path& dir)
     }
 }
 
+Manager::~Manager()
+{
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    advanced_route_worker.std::thread::~thread();
+#endif
+}
+
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+void Manager::AdvancedRoute()
+{
+    while (true)
+    {
+        advanced_route_cond_var.wait(advanced_route_lock);
+        std::this_thread::sleep_for(std::chrono::seconds(5));
+
+        for (auto it = interfaces.begin(); it != interfaces.end(); it++)
+        {
+            auto ifname = it->first;
+            execute("/usr/bin/ipv4-advanced-route.sh", "ipv4-advanced-route.sh",
+                    ifname.c_str(), it->second->linkUp() ? "UP" : "DOWN");
+
+            execute("/usr/bin/ipv6-advanced-route.sh", "ipv6-advanced-route.sh",
+                    ifname.c_str(), it->second->linkUp() ? "UP" : "DOWN");
+        }
+    }
+}
+#endif
+
 } // namespace network
 } // namespace phosphor
diff --git a/src/network_manager.hpp b/src/network_manager.hpp
index bdffc42..b818bef 100644
--- a/src/network_manager.hpp
+++ b/src/network_manager.hpp
@@ -14,10 +14,13 @@
 #include <stdplus/zstring_view.hpp>
 #include <xyz/openbmc_project/Common/FactoryReset/server.hpp>
 
+#include <condition_variable>
 #include <filesystem>
 #include <memory>
+#include <mutex>
 #include <string>
 #include <string_view>
+#include <thread>
 #include <vector>
 
 namespace phosphor
@@ -41,6 +44,7 @@ class Manager : public ManagerIface
   public:
     Manager(Manager&&) = delete;
     Manager& operator=(Manager&&) = delete;
+    ~Manager();
 
     /** @brief Constructor to put object onto bus at a dbus path.
      *  @param[in] bus - Bus to attach to.
@@ -147,6 +151,11 @@ class Manager : public ManagerIface
         reloadPostHooks.push_back(std::move(hook));
     }
 
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    /** @brief Used to notify/wait to exexute advanced-route */
+    std::condition_variable advanced_route_cond_var;
+#endif
+
   protected:
     /** @brief Handle to the object used to trigger reloads of networkd. */
     stdplus::PinnedRef<DelayedExecutor> reload;
@@ -190,6 +199,14 @@ class Manager : public ManagerIface
 
     /** @brief Creates the interface in the maps */
     void createInterface(const AllIntfInfo& info, bool enabled);
+
+  private:
+#ifdef AMI_IP_ADVANCED_ROUTING_SUPPORT
+    std::thread advanced_route_worker;
+    void AdvancedRoute();
+    std::unique_lock<std::mutex> advanced_route_lock;
+    std::mutex advanced_route_mutex;
+#endif
 };
 
 } // namespace network
-- 
2.43.0


